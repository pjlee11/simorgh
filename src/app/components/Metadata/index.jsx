import React from 'react';
import Helmet from 'react-helmet';
import { arrayOf, bool, string, number } from 'prop-types';

const Metadata = ({
  isAmp,
  articleAuthor,
  articleSection,
  brandName,
  canonicalLink,
  defaultImage,
  defaultImageAltText,
  description,
  facebookAdmin,
  facebookAppID,
  lang,
  locale,
  metaTags,
  timeFirstPublished,
  timeLastUpdated,
  title,
  twitterCreator,
  twitterSite,
  type,
}) => {
  const htmlAttributes = { lang };

  if (isAmp) {
    htmlAttributes.amp = ''; // empty value as this makes Helmet render 'amp' as per https://www.ampproject.org/docs/fundamentals/spec#ampd
  }

  const injectAmpScript = isAmp ? (
    <script key="amp" async src="https://cdn.ampproject.org/v0.js" />
  ) : null;

  return (
    <Helmet htmlAttributes={htmlAttributes}>
      <meta
        name="viewport"
        content="width=device-width, initial-scale=1, minimum-scale=1"
      />
      <title>
        {title} &#8211; {brandName}
      </title>
      <link rel="canonical" href={canonicalLink} />
      <meta name="article:author" content={articleAuthor} />
      <meta name="article:modified_time" content={timeLastUpdated} />
      <meta name="article:published_time" content={timeFirstPublished} />
      {articleSection ? (
        <meta name="article:section" content={articleSection} />
      ) : null}
      {metaTags.map(tag => (
        <meta name="article:tag" content={tag} key={tag} />
      ))}
      <meta name="description" content={description} />
      <meta name="fb:admins" content={facebookAdmin} />
      <meta name="fb:app_id" content={facebookAppID} />
      <meta name="og:description" content={description} />
      <meta name="og:image" content={defaultImage} />
      <meta name="og:image:alt" content={defaultImageAltText} />
      <meta name="og:locale" content={locale} />
      <meta name="og:site_name" content={brandName} />
      <meta name="og:title" content={title} />
      <meta name="og:type" content={type} />
      <meta name="og:url" content={canonicalLink} />
      <meta name="twitter:card" content="summary_large_image" />
      <meta name="twitter:creator" content={twitterCreator} />
      <meta name="twitter:description" content={description} />
      <meta name="twitter:image:alt" content={defaultImageAltText} />
      <meta name="twitter:image:src" content={defaultImage} />
      <meta name="twitter:site" content={twitterSite} />
      <meta name="twitter:title" content={title} />
      {injectAmpScript}
    </Helmet>
  );
};

Metadata.propTypes = {
  isAmp: bool.isRequired,
  articleAuthor: string.isRequired,
  articleSection: string,
  brandName: string.isRequired,
  canonicalLink: string.isRequired,
  defaultImage: string.isRequired,
  defaultImageAltText: string.isRequired,
  description: string.isRequired,
  facebookAdmin: number.isRequired,
  facebookAppID: number.isRequired,
  lang: string.isRequired,
  locale: string.isRequired,
  metaTags: arrayOf(string).isRequired,
  timeFirstPublished: string.isRequired,
  timeLastUpdated: string.isRequired,
  title: string,
  twitterCreator: string.isRequired,
  twitterSite: string.isRequired,
  type: string.isRequired,
};

Metadata.defaultProps = {
  articleSection: null,
  title: '', // defaulting as an empty string because Optimo cannot produce seoHeadlines at this time
};

export default Metadata;
