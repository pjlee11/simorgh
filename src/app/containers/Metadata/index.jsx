import React from 'react';
import { bool, string, shape } from 'prop-types';
import { ServiceContextConsumer } from '../../components/ServiceContext';
import Metadata from '../../components/Metadata';
import metadataPropTypes from '../../models/propTypes/metadata';
import promoPropTypes from '../../models/propTypes/promo';

/* An array of each thingLabel from tags.about & tags.mention */
const allTags = tags => {
  const { about, mentions } = tags;
  const aboutTags = about ? about.map(thing => thing.thingLabel) : [];
  const mentionTags = mentions ? mentions.map(thing => thing.thingLabel) : [];
  return aboutTags.concat(mentionTags);
};

const MetadataContainer = ({ isAmp, metadata, promo, service }) => {
  const { id: aresArticleId } = metadata;

  if (!aresArticleId) {
    return null;
  }

  const id = aresArticleId.split(':').pop();
  /* Canonical link generated from servicename and id */
  const canonicalLink = `https://www.bbc.com/${service}/articles/${id}`;
  const timeFirstPublished = new Date(metadata.firstPublished).toISOString();
  const timeLastUpdated = new Date(metadata.lastUpdated).toISOString();

  return (
    <ServiceContextConsumer>
      {({
        brandName,
        articleAuthor,
        defaultImage,
        defaultImageAltText,
        locale,
        twitterCreator,
        twitterSite,
      }) => (
        <Metadata
          isAmp={isAmp}
          articleAuthor={articleAuthor}
          articleSection={metadata.passport.genre}
          brandName={brandName}
          canonicalLink={canonicalLink}
          defaultImage={defaultImage}
          defaultImageAltText={defaultImageAltText}
          description={promo.summary}
          facebookAdmin={100004154058350}
          facebookAppID={1609039196070050}
          lang={metadata.passport.language}
          locale={locale}
          metaTags={allTags(metadata.tags)}
          timeFirstPublished={timeFirstPublished}
          timeLastUpdated={timeLastUpdated}
          title={promo.headlines.seoHeadline}
          twitterCreator={twitterCreator}
          twitterSite={twitterSite}
          type={metadata.type}
        />
      )}
    </ServiceContextConsumer>
  );
};

MetadataContainer.propTypes = {
  isAmp: bool.isRequired,
  metadata: shape(metadataPropTypes).isRequired,
  promo: shape(promoPropTypes).isRequired,
  service: string.isRequired,
};

export default MetadataContainer;
