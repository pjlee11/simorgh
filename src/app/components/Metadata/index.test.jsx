import React from 'react';
import Metadata from './index';
import { shouldShallowMatchSnapshot } from '../../helpers/tests/testHelpers';

const metadataSnapshotTest = (
  testDescription,
  isAmp,
  articleAuthor,
  articleSection,
  brandName,
  canonicalLink,
  defaultImage,
  defaultImageAltText,
  description,
  facebookAdmin,
  facebookAppID,
  lang,
  locale,
  metaTags,
  timeLastUpdated,
  timeFirstPublished,
  title,
  twitterCreator,
  twitterSite,
) =>
  describe(testDescription, () => {
    const metadataProps = {
      isAmp,
      articleAuthor,
      articleSection,
      brandName,
      canonicalLink,
      defaultImage,
      defaultImageAltText,
      description,
      facebookAdmin,
      facebookAppID,
      lang,
      locale,
      metaTags,
      timeLastUpdated,
      timeFirstPublished,
      title,
      twitterCreator,
      twitterSite,
    };

    shouldShallowMatchSnapshot(
      'should render correctly',
      <Metadata {...metadataProps} />,
    );
  });

describe('Metadata', () => {
  metadataSnapshotTest(
    'News article',
    false,
    'BBC News',
    null,
    'BBC News',
    'https://www.bbc.com/news/articles/c0000000001o',
    'https://www.bbc.com/news/image.png',
    'BBC News',
    'This is a description',
    101010,
    202020,
    'en-GB',
    'en_GB',
    ['tagA', 'tagB'],
    1514811600000,
    1514811600000,
    'An article title',
    '@BBCNews',
    '@BBCNews',
  );

  metadataSnapshotTest(
    'News article with null title value',
    false,
    'BBC News',
    null,
    'BBC News',
    'https://www.bbc.com/news/articles/c0000000001o',
    'https://www.bbc.com/news/image.png',
    'BBC News',
    'This is a description',
    101010,
    202020,
    'en-GB',
    'en_GB',
    ['tagA', 'tagB'],
    1514811600000,
    1514811600000,
    null,
    '@BBCNews',
    '@BBCNews',
  );

  metadataSnapshotTest(
    'News AMP article',
    true,
    'BBC News',
    null,
    'BBC News',
    'https://www.bbc.com/news/articles/c0000000001o.amp',
    'https://www.bbc.com/news/image.png',
    'BBC News',
    'This is a description',
    101010,
    202020,
    'en-GB',
    'en_GB',
    ['tagA', 'tagB'],
    1514811600000,
    1514811600000,
    'An article title',
    '@BBCNews',
    '@BBCNews',
  );

  metadataSnapshotTest(
    'Persian article',
    false,
    'BBC News فارسی',
    null,
    'BBC News فارسی',
    'https://www.bbc.com/persian/articles/cwv2xv848j5o',
    'https://www.bbc.com/persian/image.png',
    'BBC News فارسی',
    'This is a description',
    101010,
    202020,
    'fa',
    'fa',
    ['tagA', 'tagB'],
    1514811600000,
    1514811600000,
    'پهپادی که برایتان قهوه می‌آورد',
    '@bbcpersian',
    '@bbcpersian',
  );

  metadataSnapshotTest(
    'Persian AMP article',
    true,
    'BBC News فارسی',
    null,
    'BBC News فارسی',
    'https://www.bbc.com/persian/articles/cwv2xv848j5o.amp',
    'https://www.bbc.com/persian/image.png',
    'BBC News فارسی',
    'This is a description',
    101010,
    202020,
    'fa',
    'fa',
    ['tagA', 'tagB'],
    1514811600000,
    1514811600000,
    'پهپادی که برایتان قهوه می‌آورد',
    '@bbcpersian',
    '@bbcpersian',
  );

  metadataSnapshotTest(
    'articleSection is not null',
    false,
    'BBC News',
    'Politics',
    'BBC News',
    'https://www.bbc.com/news/articles/c0000000001o',
    'https://www.bbc.com/news/image.png',
    'BBC News',
    'This is a description',
    101010,
    202020,
    'en-GB',
    'en_GB',
    ['tagA', 'tagB'],
    1514811600000,
    1514811600000,
    'An article title',
    '@BBCNews',
    '@BBCNews',
  );
});
