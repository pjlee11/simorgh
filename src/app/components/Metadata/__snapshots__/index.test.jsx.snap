// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Metadata News AMP article should render correctly 1`] = `
<HelmetWrapper
  defer={true}
  encodeSpecialCharacters={true}
  htmlAttributes={
    Object {
      "amp": "",
      "lang": "en-GB",
    }
  }
>
  <meta
    content="width=device-width, initial-scale=1, minimum-scale=1"
    name="viewport"
  />
  <title>
    An article title
     – 
    BBC News
  </title>
  <link
    href="https://www.bbc.com/news/articles/c0000000001o.amp"
    rel="canonical"
  />
  <meta
    content="BBC News"
    name="article:author"
  />
  <meta
    content={1514811600000}
    name="article:modified_time"
  />
  <meta
    content={1514811600000}
    name="article:published_time"
  />
  <meta
    content="tagA"
    name="article:tag"
  />
  <meta
    content="tagB"
    name="article:tag"
  />
  <meta
    content="This is a description"
    name="description"
  />
  <meta
    content={101010}
    name="fb:admins"
  />
  <meta
    content={202020}
    name="fb:app_id"
  />
  <meta
    content="This is a description"
    name="og:description"
  />
  <meta
    content="https://www.bbc.com/news/image.png"
    name="og:image"
  />
  <meta
    content="BBC News"
    name="og:image:alt"
  />
  <meta
    content="en_GB"
    name="og:locale"
  />
  <meta
    content="BBC News"
    name="og:site_name"
  />
  <meta
    content="An article title"
    name="og:title"
  />
  <meta
    name="og:type"
  />
  <meta
    content="https://www.bbc.com/news/articles/c0000000001o.amp"
    name="og:url"
  />
  <meta
    content="summary_large_image"
    name="twitter:card"
  />
  <meta
    content="@BBCNews"
    name="twitter:creator"
  />
  <meta
    content="This is a description"
    name="twitter:description"
  />
  <meta
    content="BBC News"
    name="twitter:image:alt"
  />
  <meta
    content="https://www.bbc.com/news/image.png"
    name="twitter:image:src"
  />
  <meta
    content="@BBCNews"
    name="twitter:site"
  />
  <meta
    content="An article title"
    name="twitter:title"
  />
  <script
    async={true}
    src="https://cdn.ampproject.org/v0.js"
  />
</HelmetWrapper>
`;

exports[`Metadata News article should render correctly 1`] = `
<HelmetWrapper
  defer={true}
  encodeSpecialCharacters={true}
  htmlAttributes={
    Object {
      "lang": "en-GB",
    }
  }
>
  <meta
    content="width=device-width, initial-scale=1, minimum-scale=1"
    name="viewport"
  />
  <title>
    An article title
     – 
    BBC News
  </title>
  <link
    href="https://www.bbc.com/news/articles/c0000000001o"
    rel="canonical"
  />
  <meta
    content="BBC News"
    name="article:author"
  />
  <meta
    content={1514811600000}
    name="article:modified_time"
  />
  <meta
    content={1514811600000}
    name="article:published_time"
  />
  <meta
    content="tagA"
    name="article:tag"
  />
  <meta
    content="tagB"
    name="article:tag"
  />
  <meta
    content="This is a description"
    name="description"
  />
  <meta
    content={101010}
    name="fb:admins"
  />
  <meta
    content={202020}
    name="fb:app_id"
  />
  <meta
    content="This is a description"
    name="og:description"
  />
  <meta
    content="https://www.bbc.com/news/image.png"
    name="og:image"
  />
  <meta
    content="BBC News"
    name="og:image:alt"
  />
  <meta
    content="en_GB"
    name="og:locale"
  />
  <meta
    content="BBC News"
    name="og:site_name"
  />
  <meta
    content="An article title"
    name="og:title"
  />
  <meta
    name="og:type"
  />
  <meta
    content="https://www.bbc.com/news/articles/c0000000001o"
    name="og:url"
  />
  <meta
    content="summary_large_image"
    name="twitter:card"
  />
  <meta
    content="@BBCNews"
    name="twitter:creator"
  />
  <meta
    content="This is a description"
    name="twitter:description"
  />
  <meta
    content="BBC News"
    name="twitter:image:alt"
  />
  <meta
    content="https://www.bbc.com/news/image.png"
    name="twitter:image:src"
  />
  <meta
    content="@BBCNews"
    name="twitter:site"
  />
  <meta
    content="An article title"
    name="twitter:title"
  />
</HelmetWrapper>
`;

exports[`Metadata News article with null title value should render correctly 1`] = `
<HelmetWrapper
  defer={true}
  encodeSpecialCharacters={true}
  htmlAttributes={
    Object {
      "lang": "en-GB",
    }
  }
>
  <meta
    content="width=device-width, initial-scale=1, minimum-scale=1"
    name="viewport"
  />
  <title>
     – 
    BBC News
  </title>
  <link
    href="https://www.bbc.com/news/articles/c0000000001o"
    rel="canonical"
  />
  <meta
    content="BBC News"
    name="article:author"
  />
  <meta
    content={1514811600000}
    name="article:modified_time"
  />
  <meta
    content={1514811600000}
    name="article:published_time"
  />
  <meta
    content="tagA"
    name="article:tag"
  />
  <meta
    content="tagB"
    name="article:tag"
  />
  <meta
    content="This is a description"
    name="description"
  />
  <meta
    content={101010}
    name="fb:admins"
  />
  <meta
    content={202020}
    name="fb:app_id"
  />
  <meta
    content="This is a description"
    name="og:description"
  />
  <meta
    content="https://www.bbc.com/news/image.png"
    name="og:image"
  />
  <meta
    content="BBC News"
    name="og:image:alt"
  />
  <meta
    content="en_GB"
    name="og:locale"
  />
  <meta
    content="BBC News"
    name="og:site_name"
  />
  <meta
    content={null}
    name="og:title"
  />
  <meta
    name="og:type"
  />
  <meta
    content="https://www.bbc.com/news/articles/c0000000001o"
    name="og:url"
  />
  <meta
    content="summary_large_image"
    name="twitter:card"
  />
  <meta
    content="@BBCNews"
    name="twitter:creator"
  />
  <meta
    content="This is a description"
    name="twitter:description"
  />
  <meta
    content="BBC News"
    name="twitter:image:alt"
  />
  <meta
    content="https://www.bbc.com/news/image.png"
    name="twitter:image:src"
  />
  <meta
    content="@BBCNews"
    name="twitter:site"
  />
  <meta
    content={null}
    name="twitter:title"
  />
</HelmetWrapper>
`;

exports[`Metadata Persian AMP article should render correctly 1`] = `
<HelmetWrapper
  defer={true}
  encodeSpecialCharacters={true}
  htmlAttributes={
    Object {
      "amp": "",
      "lang": "fa",
    }
  }
>
  <meta
    content="width=device-width, initial-scale=1, minimum-scale=1"
    name="viewport"
  />
  <title>
    پهپادی که برایتان قهوه می‌آورد
     – 
    BBC News فارسی
  </title>
  <link
    href="https://www.bbc.com/persian/articles/cwv2xv848j5o.amp"
    rel="canonical"
  />
  <meta
    content="BBC News فارسی"
    name="article:author"
  />
  <meta
    content={1514811600000}
    name="article:modified_time"
  />
  <meta
    content={1514811600000}
    name="article:published_time"
  />
  <meta
    content="tagA"
    name="article:tag"
  />
  <meta
    content="tagB"
    name="article:tag"
  />
  <meta
    content="This is a description"
    name="description"
  />
  <meta
    content={101010}
    name="fb:admins"
  />
  <meta
    content={202020}
    name="fb:app_id"
  />
  <meta
    content="This is a description"
    name="og:description"
  />
  <meta
    content="https://www.bbc.com/persian/image.png"
    name="og:image"
  />
  <meta
    content="BBC News فارسی"
    name="og:image:alt"
  />
  <meta
    content="fa"
    name="og:locale"
  />
  <meta
    content="BBC News فارسی"
    name="og:site_name"
  />
  <meta
    content="پهپادی که برایتان قهوه می‌آورد"
    name="og:title"
  />
  <meta
    name="og:type"
  />
  <meta
    content="https://www.bbc.com/persian/articles/cwv2xv848j5o.amp"
    name="og:url"
  />
  <meta
    content="summary_large_image"
    name="twitter:card"
  />
  <meta
    content="@bbcpersian"
    name="twitter:creator"
  />
  <meta
    content="This is a description"
    name="twitter:description"
  />
  <meta
    content="BBC News فارسی"
    name="twitter:image:alt"
  />
  <meta
    content="https://www.bbc.com/persian/image.png"
    name="twitter:image:src"
  />
  <meta
    content="@bbcpersian"
    name="twitter:site"
  />
  <meta
    content="پهپادی که برایتان قهوه می‌آورد"
    name="twitter:title"
  />
  <script
    async={true}
    src="https://cdn.ampproject.org/v0.js"
  />
</HelmetWrapper>
`;

exports[`Metadata Persian article should render correctly 1`] = `
<HelmetWrapper
  defer={true}
  encodeSpecialCharacters={true}
  htmlAttributes={
    Object {
      "lang": "fa",
    }
  }
>
  <meta
    content="width=device-width, initial-scale=1, minimum-scale=1"
    name="viewport"
  />
  <title>
    پهپادی که برایتان قهوه می‌آورد
     – 
    BBC News فارسی
  </title>
  <link
    href="https://www.bbc.com/persian/articles/cwv2xv848j5o"
    rel="canonical"
  />
  <meta
    content="BBC News فارسی"
    name="article:author"
  />
  <meta
    content={1514811600000}
    name="article:modified_time"
  />
  <meta
    content={1514811600000}
    name="article:published_time"
  />
  <meta
    content="tagA"
    name="article:tag"
  />
  <meta
    content="tagB"
    name="article:tag"
  />
  <meta
    content="This is a description"
    name="description"
  />
  <meta
    content={101010}
    name="fb:admins"
  />
  <meta
    content={202020}
    name="fb:app_id"
  />
  <meta
    content="This is a description"
    name="og:description"
  />
  <meta
    content="https://www.bbc.com/persian/image.png"
    name="og:image"
  />
  <meta
    content="BBC News فارسی"
    name="og:image:alt"
  />
  <meta
    content="fa"
    name="og:locale"
  />
  <meta
    content="BBC News فارسی"
    name="og:site_name"
  />
  <meta
    content="پهپادی که برایتان قهوه می‌آورد"
    name="og:title"
  />
  <meta
    name="og:type"
  />
  <meta
    content="https://www.bbc.com/persian/articles/cwv2xv848j5o"
    name="og:url"
  />
  <meta
    content="summary_large_image"
    name="twitter:card"
  />
  <meta
    content="@bbcpersian"
    name="twitter:creator"
  />
  <meta
    content="This is a description"
    name="twitter:description"
  />
  <meta
    content="BBC News فارسی"
    name="twitter:image:alt"
  />
  <meta
    content="https://www.bbc.com/persian/image.png"
    name="twitter:image:src"
  />
  <meta
    content="@bbcpersian"
    name="twitter:site"
  />
  <meta
    content="پهپادی که برایتان قهوه می‌آورد"
    name="twitter:title"
  />
</HelmetWrapper>
`;

exports[`Metadata articleSection is not null should render correctly 1`] = `
<HelmetWrapper
  defer={true}
  encodeSpecialCharacters={true}
  htmlAttributes={
    Object {
      "lang": "en-GB",
    }
  }
>
  <meta
    content="width=device-width, initial-scale=1, minimum-scale=1"
    name="viewport"
  />
  <title>
    An article title
     – 
    BBC News
  </title>
  <link
    href="https://www.bbc.com/news/articles/c0000000001o"
    rel="canonical"
  />
  <meta
    content="BBC News"
    name="article:author"
  />
  <meta
    content={1514811600000}
    name="article:modified_time"
  />
  <meta
    content={1514811600000}
    name="article:published_time"
  />
  <meta
    content="Politics"
    name="article:section"
  />
  <meta
    content="tagA"
    name="article:tag"
  />
  <meta
    content="tagB"
    name="article:tag"
  />
  <meta
    content="This is a description"
    name="description"
  />
  <meta
    content={101010}
    name="fb:admins"
  />
  <meta
    content={202020}
    name="fb:app_id"
  />
  <meta
    content="This is a description"
    name="og:description"
  />
  <meta
    content="https://www.bbc.com/news/image.png"
    name="og:image"
  />
  <meta
    content="BBC News"
    name="og:image:alt"
  />
  <meta
    content="en_GB"
    name="og:locale"
  />
  <meta
    content="BBC News"
    name="og:site_name"
  />
  <meta
    content="An article title"
    name="og:title"
  />
  <meta
    name="og:type"
  />
  <meta
    content="https://www.bbc.com/news/articles/c0000000001o"
    name="og:url"
  />
  <meta
    content="summary_large_image"
    name="twitter:card"
  />
  <meta
    content="@BBCNews"
    name="twitter:creator"
  />
  <meta
    content="This is a description"
    name="twitter:description"
  />
  <meta
    content="BBC News"
    name="twitter:image:alt"
  />
  <meta
    content="https://www.bbc.com/news/image.png"
    name="twitter:image:src"
  />
  <meta
    content="@BBCNews"
    name="twitter:site"
  />
  <meta
    content="An article title"
    name="twitter:title"
  />
</HelmetWrapper>
`;
